### ISSUE TITLE
Multiple projects in a single repo

### ISSUE THREAD

**Issue by NickeZ** [borkdude/clj-kondo#1051]
**Is your feature request related to a problem? Please describe.**
I have many projects in a single repo. The pre-commit hook suggested in the documentation does not pick up on the `.clj-kondo` directory in the project directory.

**Describe the solution you'd like**
It would be great if clj-kondo could pick up the correct config file for the linted file based on the first `.clj-kondo` directory it finds while traversing upwards.

**Describe alternatives you've considered**
Otherwise I'll have to write a more complex pre-commit hook that searches for the `.clj-kondo` dir per file.

**Comment by borkdude** [comment]
> It would be great if clj-kondo could pick up the correct config file for the linted file based on the first .clj-kondo directory it finds while traversing upwards.

It does so, based on the current working directory. Alternatively you can change this by setting `--config-dir` manually to `$project/repo/.clj-kondo`.

**Comment by NickeZ** [comment]
```sh
if !(git diff --cached --name-only --diff-filter=AM $against | grep -E '.clj[cs]?$' | xargs clj-kondo --lint)
then
    echo
    echo "Error: new clj-kondo errors found. Please fix them and retry the commit."
    exit 1
fi
```

I'd have to make this a two-pass thing. First list files, then find config-dirs, and then pass the correct config dir to the corresponding file.

**Comment by borkdude** [comment]
@NickeZ How many sub-projects do you have? Maybe you could iterate over them in a for loop while setting the cwd to each project and lint the projects separately?

**Comment by NickeZ** [comment]
Not too many, but we use a monorepo and it won't decrease over time I suppose..

that sounds more complicated, I still need to match the right project to the right clojure file. I still only want to check the changed files.

**Comment by borkdude** [comment]
@NickeZ I really recommend using babashka (in favor of bash) for this. I created this script while we are discussing this issue, just to give you an idea how much easier this is in Clojure:

``` clojure
#!/usr/bin/env bb

(require '[babashka.pods :as pods]
         '[clojure.java.io :as io])

(pods/load-pod "clj-kondo")

(require '[pod.borkdude.clj-kondo :as clj-kondo])

(def projects ["app" "annotator"])

(def results
  (reduce (fn [acc proj]
            (merge acc (clj-kondo/run! {:lint [(.getPath (io/file proj "src"))]
                                        :config-dir (.getPath (io/file proj ".clj-kondo"))})))
          {}
          projects))

(prn (:summary results))

(defn exit-code [results]
  (if (pos? (-> results :summary :error))
    1
    0))

(System/exit (exit-code results))
```

``` clojure
$ bb lint.clj
{:error 114, :warning 818, :info 0, :type :summary, :duration 1136}
borkdude@MBP2019 ~/dre/DocSearch (tags4*) $ echo $?
1
```

**Comment by NickeZ** [comment]
I'll take a look at bb, obviously clojure is easier to script in than bash. But here is my attempt so far anyway..

```sh
files=$(git diff --cached --name-only --diff-filter=AM $against | grep -E '.clj[cs]?$' )
for file in $files; do
    x=`dirname $file`
    while [ "$x" != "/" ] ; do
        x=`dirname "$x"`
        if stat $x/.clj-kondo > /dev/null; then
            kondodir=$x/.clj-kondo
        fi 
    done
    clj-kondo --lint $file --config-dir $kondodir
done
```

**Comment by NickeZ** [comment]
is this babashka the only way to write single-file clojure scripts?

**Comment by borkdude** [comment]
(Note: in the above script `merge` is probably not the right way to do it, it needs a little bit more sophistication)

**Comment by borkdude** [comment]
@NickeZ No, you can also use Clojure on the JVM to write scripts. The benefit of babashka is instant startup time whereas the JVM takes one to a few seconds (with Clojure anyway). You could also write python or whatever you like, of course.

**Comment by borkdude** [comment]
Updated the above script to fix merging of multiple project results.

**Comment by borkdude** [comment]
The way I'm going this myself in a multiple project repo:

```
# lint app

cd "$PWD/app"

if !(git diff --relative --cached --name-only --diff-filter=AM $against | grep -E '.clj[cs]?$' | xargs clj-kondo --lint)
then
    echo
    echo "Error: new clj-kondo errors found. Please fix them and retry the commit."
    exit 1
fi
```
for each project.

**Comment by NickeZ** [comment]
You can close the ticket in case you won't implement this, I still think that clj-kondo could do the right thing by default here. But that is of course only my opinion.

**Comment by borkdude** [comment]
@NickeZ For your commit hook the proposed behaviour makes sense but in other cases this might not be the right behaviour. E.g. when linting a classpath, the config should not be looked up by the linted file, but using the current working directory.
